tchannel/peer_heap.py

```diff
@@ -80,3 +80,15 @@ def peek_peer(self):
             return None
 
         return self.peers[0]
+
+    def remove_peer(self, peer):
+        """Remove the peer from the heap.
+        Note: Unknown peers are ignored.
+
+        Return: removed peer if peer exists, Otherwise return None.
+        """
+        if 0 <= peer.index < self.size():
+            assert (peer is self.peers[peer.index],
+                    "peer doesn't exist in the heap")
+            p = heap.remove(self, peer.index)
+            return p
```

Comment:
I know you said that unknown peer index is a no-op but there seems to be a discrepancy in behavior here.

- If your peer is unknown AND it has an index outside our range, we'll do nothing.
- If your peer is unknown but its index is within our range, we'll raise an exception.

I think we should raise an exception in both cases. In the second case, you can raise an IndexError for the out-of-range case (the one where it's a no-op).